Remaining test fails under django2.2

https://travis-ci.org/mozilla/addons-server/jobs/498900845#L4465
**________________ TestAddonManager.test_filter_for_many_to_many _________________**
```
src/olympia/addons/tests/test_models.py:345: in test_filter_for_many_to_many
    assert collection.addons.count() == 0
E   AssertionError: assert 1 == 0
```

~https://travis-ci.org/mozilla/addons-server/jobs/498900846#L3695
**______________ TestEditMedia.test_edit_media_shows_proper_labels _______________**
/devhub/tests/test_views_edit.py:888: in test_edit_media_shows_proper_labels~

https://travis-ci.org/mozilla/addons-server/jobs/498900846#L3702
**_ TestAddonSubmitAgreementWithPostReviewEnabled.test_read_dev_agreement_captcha_active_error _**
```
src/olympia/devhub/tests/test_views_submit.py:228: in test_read_dev_agreement_captcha_active_error
    response = self.client.get(reverse('devhub.submit.agreement'))
.tox/devhub/lib/python3.6/site-packages/django/test/client.py:535: in get
    response = super().get(path, data=data, secure=secure, **extra)
.tox/devhub/lib/python3.6/site-packages/django/test/client.py:347: in get
    **extra,
.tox/devhub/lib/python3.6/site-packages/django/test/client.py:422: in generic
    return self.request(**r)
.tox/devhub/lib/python3.6/site-packages/django/test/client.py:503: in request
    raise exc_value
.tox/devhub/lib/python3.6/site-packages/django/core/handlers/exception.py:34: in inner
    response = get_response(request)
.tox/devhub/lib/python3.6/site-packages/django/core/handlers/base.py:115: in _get_response
    response = self.process_exception_by_middleware(e, request)
.tox/devhub/lib/python3.6/site-packages/django/core/handlers/base.py:113: in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
/opt/python/3.6.7/lib/python3.6/contextlib.py:52: in inner
    return func(*args, **kwds)
src/olympia/amo/decorators.py:30: in wrapper
    return func(request, *args, **kw)
src/olympia/devhub/views.py:1199: in submit_addon
    render_captcha=waffle.switch_is_active('addon-submission-captcha')
src/olympia/devhub/views.py:1742: in render_agreement
    return render(request, template, context)
src/olympia/amo/utils.py:68: in render
    rendered = loader.render_to_string(template, ctx, request=request)
.tox/devhub/lib/python3.6/site-packages/django/template/loader.py:62: in render_to_string
    return template.render(context, request)
.tox/devhub/lib/python3.6/site-packages/django_jinja/backend.py:106: in render
    return mark_safe(self.template.render(context))
conftest.py:114: in instrumented_render
    return old_render(self, *args, **kwargs)
.tox/devhub/lib/python3.6/site-packages/jinja2/asyncsupport.py:76: in render
    return original_render(self, *args, **kwargs)
.tox/devhub/lib/python3.6/site-packages/jinja2/environment.py:1008: in render
    return self.environment.handle_exception(exc_info, True)
.tox/devhub/lib/python3.6/site-packages/jinja2/environment.py:780: in handle_exception
    reraise(exc_type, exc_value, tb)
.tox/devhub/lib/python3.6/site-packages/jinja2/_compat.py:37: in reraise
    raise value.with_traceback(tb)
src/olympia/devhub/templates/devhub/addons/submit/start.html:1: in top-level template code
    {% extends "devhub/addons/submit/base.html" %}
src/olympia/devhub/templates/devhub/addons/submit/base.html:6: in top-level template code
    {% set title = _('Submit a New Add-on') %}
src/olympia/devhub/templates/devhub/base.html:45: in top-level template code
    {% set hide_mobile_link=True %}
src/olympia/templates/base.html:133: in top-level template code
    {% block main_content %}
src/olympia/templates/base.html:136: in block "main_content"
    {% block content %}{% endblock %}
src/olympia/devhub/templates/devhub/addons/submit/base.html:25: in block "content"
    {% block primary %}{% endblock %}
src/olympia/devhub/templates/devhub/addons/submit/start.html:9: in block "primary"
    {% include "devhub/agreement.html" %}
src/olympia/devhub/templates/devhub/agreement.html:20: in top-level template code
    {{ agreement_form.recaptcha }}
.tox/devhub/lib/python3.6/site-packages/django/utils/html.py:389: in <lambda>
    klass.__html__ = lambda self: str(self)
.tox/devhub/lib/python3.6/site-packages/django/utils/html.py:388: in <lambda>
    klass.__str__ = lambda self: mark_safe(klass_str(self))
.tox/devhub/lib/python3.6/site-packages/django/forms/boundfield.py:33: in __str__
    return self.as_widget()
.tox/devhub/lib/python3.6/site-packages/django/forms/boundfield.py:93: in as_widget
    renderer=self.form.renderer,
E   TypeError: render() got an unexpected keyword argument 'renderer'
```

**_ TestAddonSubmitAgreementWithPostReviewEnabled.test_read_dev_agreement_captcha_active_success _**
```
src/olympia/devhub/tests/test_views_submit.py:244: in test_read_dev_agreement_captcha_active_success
(same as last traceback)
```

~https://travis-ci.org/mozilla/addons-server/jobs/498900846#L3836
**_____________________ TestAddonSubmitSource.test_no_source _____________________**
src/olympia/devhub/tests/test_views_submit.py:830: in test_no_source
(same exception repeated in a number of following tests)~

https://travis-ci.org/mozilla/addons-server/jobs/499409208#L3698
**_________________________ TestDelete.test_post_lwtheme _________________________**
```
src/olympia/devhub/tests/test_views.py:479: in test_post_lwtheme
    assert pq(response.content)('.notification-box').text() == (
E   AssertionError: assert '<h2> Theme deleted. </h2>' == 'Theme deleted.'
E     - <h2> Theme deleted. </h2>
E     + Theme deleted.
```
( + 6 more in TestDelete )

https://travis-ci.org/mozilla/addons-server/jobs/499296540#L3831
**_ TestVersionSubmitUploadListed.test_same_version_if_previous_is_awaiting_review _**
```
src/olympia/devhub/tests/test_views_submit.py:1924: in test_same_version_if_previous_is_awaiting_review
    assert pq(response.content)('ul.errorlist').text() == (
E   AssertionError: assert 'Version 0.1 ... instead?</a>' == 'Version 0.1 a...load instead?'
E     - Version 0.1 already exists. <a href="/en-US/developers/addon/a3615/versions/submit/81551/details">Continue with existing upload instead?</a>
E     + Version 0.1 already exists. Continue with existing upload instead?
```
(it's essentially the same problem as the TestDelete test fails - html is being escaped and returned as text)

https://travis-ci.org/mozilla/addons-server/jobs/499296540#L3838
**_ TestVersionSubmitUploadUnlisted.test_same_version_if_previous_is_awaiting_review _**
```
src/olympia/devhub/tests/test_views_submit.py:1924: in test_same_version_if_previous_is_awaiting_review
(same as last traceback)
```

taking them one at a time:
https://travis-ci.org/mozilla/addons-server/jobs/498900845#L4465
**________________ TestAddonManager.test_filter_for_many_to_many _________________**

```
src/olympia/addons/tests/test_models.py:345: in test_filter_for_many_to_many
    assert collection.addons.count() == 0
E   AssertionError: assert 1 == 0
```
Needs further investigation - it could be something trivial like needing a `Collection.reload()`; it could be a significant problem in the way we load models and related objects.

https://travis-ci.org/mozilla/addons-server/jobs/499409208#L3698
**_________________________ TestDelete.test_post_lwtheme _________________________**

```
src/olympia/devhub/tests/test_views.py:479: in test_post_lwtheme
    assert pq(response.content)('.notification-box').text() == (
E   AssertionError: assert '<h2> Theme deleted. </h2>' == 'Theme deleted.'
E     - <h2> Theme deleted. </h2>
E     + Theme deleted.
``` 
* 7

Initially I thought it was a bug with pquery but now it seems the html content such as `<h2>` is being escaped by `django.contrib.messages` (which it shouldn't as we go through the effort of marking it safe).

https://travis-ci.org/mozilla/addons-server/jobs/499296540#L3831
**_ TestVersionSubmitUploadListed.test_same_version_if_previous_is_awaiting_review _** and **_ TestVersionSubmitUploadUnlisted.test_same_version_if_previous_is_awaiting_review _**
```
src/olympia/devhub/tests/test_views_submit.py:1924: in test_same_version_if_previous_is_awaiting_review
    assert pq(response.content)('ul.errorlist').text() == (
E   AssertionError: assert 'Version 0.1 ... instead?</a>' == 'Version 0.1 a...load instead?'
E     - Version 0.1 already exists. <a href="/en-US/developers/addon/a3615/versions/submit/81551/details">Continue with existing upload instead?</a>
E     + Version 0.1 already exists. Continue with existing upload instead?
```

Similar to the `TestDelete` test fails - html is being escaped and returned as text - but this time the message is passed via a `forms.ValidationError` in `NewUploadForm.clean()`

--- a/src/olympia/access/models.py
+++ b/src/olympia/access/models.py
@@ -6,7 +6,7 @@ from django.utils.encoding import python_2_unicode_compatible
 import olympia.core.logger
 
 from olympia import activity, amo
-from olympia.amo.fields import PositiveAutoField
+from olympia.amo.fields import ManyToManyField, PositiveAutoField
 from olympia.amo.models import ModelBase
 
 
@@ -19,8 +19,8 @@ class Group(ModelBase):
     id = PositiveAutoField(primary_key=True)
     name = models.CharField(max_length=255, default='')
     rules = models.TextField()
-    users = models.ManyToManyField('users.UserProfile', through='GroupUser',
-                                   related_name='groups')
+    users = ManyToManyField(
+        'users.UserProfile', through='GroupUser', related_name='groups')
     notes = models.TextField(blank=True)
 
     class Meta:

--- a/src/olympia/activity/models.py
+++ b/src/olympia/activity/models.py
@@ -285,7 +285,7 @@ class ActivityLog(ModelBase):
         [(value.id, key)
          for key, value in constants.activity.LOG_BY_ID.items()])
     user = models.ForeignKey(
-        'users.UserProfile', null=True, on_delete=models.CASCADE)
+        'users.UserProfile', null=True, on_delete=models.SET_NULL)
     action = models.SmallIntegerField(choices=TYPES, db_index=True)
     _arguments = models.TextField(blank=True, db_column='arguments')
     _details = models.TextField(blank=True, db_column='details')

--- a/src/olympia/addons/models.py
+++ b/src/olympia/addons/models.py
@@ -34,7 +34,7 @@ from olympia.access import acl
 from olympia.addons.utils import (
     generate_addon_guid, get_creatured_ids, get_featured_ids)
 from olympia.amo.decorators import use_primary_db
-from olympia.amo.fields import PositiveAutoField
+from olympia.amo.fields import ManyToManyField, PositiveAutoField
 from olympia.amo.models import (
     BasePreview, BaseQuerySet, ManagerBase, ModelBase, OnChangeMixin,
     SaveUpdateMixin, SlugField, manual_order)
@@ -334,9 +334,9 @@ class Addon(OnChangeMixin, ModelBase):
 
     contributions = models.URLField(max_length=255, blank=True)
 
-    authors = models.ManyToManyField('users.UserProfile', through='AddonUser',
-                                     related_name='addons')
-    categories = models.ManyToManyField('Category', through='AddonCategory')
+    authors = ManyToManyField(
+        'users.UserProfile', through='AddonUser', related_name='addons')
+    categories = ManyToManyField('Category', through='AddonCategory')
 
     _current_version = models.ForeignKey(Version, db_column='current_version',
                                          related_name='+', null=True,
@@ -1902,7 +1902,7 @@ class Category(OnChangeMixin, ModelBase):
         default=0, help_text='Category weight used in sort ordering')
     misc = models.BooleanField(default=False)
 
-    addons = models.ManyToManyField(Addon, through='AddonCategory')
+    addons = ManyToManyField(Addon, through='AddonCategory')
 
     class Meta:
         db_table = 'categories'

--- a/src/olympia/addons/tests/test_buttons.py
+++ b/src/olympia/addons/tests/test_buttons.py
@@ -2,7 +2,10 @@ import json
 
 from datetime import datetime
 
+import django
+
 import jinja2
+import pytest
 
 from mock import Mock, patch
 from pyquery import PyQuery
@@ -84,6 +87,9 @@ class ButtonTest(TestCase):
         return file
 
 
+@pytest.mark.skipif(
+    django.VERSION[0] >= 2,
+    reason='Legacy UI tests')
 class TestButtonSetup(ButtonTest):
     """Tests for setup code inside install_button."""
 
@@ -140,6 +146,9 @@ class TestButtonSetup(ButtonTest):
         assert b.version == self.version
 
 
+@pytest.mark.skipif(
+    django.VERSION[0] >= 2,
+    reason='Legacy UI tests')
 class TestButton(ButtonTest):
     """Tests for the InstallButton class."""
 
@@ -296,6 +305,9 @@ class TestButton(ButtonTest):
         }
 
 
+@pytest.mark.skipif(
+    django.VERSION[0] >= 2,
+    reason='Legacy UI tests')
 class TestButtonHtml(ButtonTest):
 
     def test_basics(self):
@@ -457,6 +469,9 @@ class TestButtonHtml(ButtonTest):
         assert doc('.download-anyway')
 
 
+@pytest.mark.skipif(
+    django.VERSION[0] >= 2,
+    reason='Legacy UI tests')
 class TestViews(TestCase):
     fixtures = ['addons/eula+contrib-addon']
 

--- a/src/olympia/addons/tests/test_views.py
+++ b/src/olympia/addons/tests/test_views.py
@@ -3,6 +3,7 @@ import json
 import random
 import re
 
+import django
 from django.conf import settings
 from django.core import mail
 from django.test.client import Client
@@ -1291,6 +1292,9 @@ class TestPrivacyPolicy(TestCase):
         check_cat_sidebar(self.url, self.addon)
 
 
+@pytest.mark.skipif(
+    django.VERSION[0] >= 2,
+    reason='Legacy UI tests')
 @patch.object(settings, 'NOBOT_RECAPTCHA_PRIVATE_KEY', 'something')
 class TestReportAbuse(TestCase):
     fixtures = ['addons/persona', 'base/addon_3615', 'base/users']

--- a/src/olympia/addons/widgets.py
+++ b/src/olympia/addons/widgets.py
@@ -15,7 +15,7 @@ class IconTypeSelect(forms.RadioSelect):
         '<a href="#" class="{active}">'
         '<img src="{static}img/addon-icons/{icon_name}-32.png" alt="">'
         '</a>'
-        '<label for="{label_id}">{original_widget}</label>'
+        '{original_widget}'
         '</li>'
     )
 
@@ -34,14 +34,20 @@ class IconTypeSelect(forms.RadioSelect):
             if option_value.split('/')[0] == 'icon' or option_value == '':
                 icon_name = option['label']
 
+                original_widget = self._render(
+                    self.option_template_name, option)
+                if not original_widget.startswith('<label for='):
+                    # django2.2+ already wraps widget in a label
+                    original_widget = format_html(
+                        '<label for="{label_id}">{widget}</label>',
+                        label_id=option['widget']['attrs']['id'],
+                        widget=original_widget)
                 output.append(format_html(
                     self.base_html,
                     active='active' if option_value == value else '',
                     static=settings.STATIC_URL,
                     icon_name=icon_name,
-                    label_id=option['widget']['attrs']['id'],
-                    original_widget=self._render(
-                        self.option_template_name, option)
+                    original_widget=original_widget,
                 ))
             else:
                 output.append(format_html(

--- a/src/olympia/amo/fields.py
+++ b/src/olympia/amo/fields.py
@@ -4,8 +4,10 @@ from django.conf import settings
 from django.core import exceptions
 from django.core.validators import RegexValidator, URLValidator
 from django.db import models
+from django.db.models.fields import related_descriptors
 from django.forms import fields
 from django.utils.translation import ugettext_lazy as _
+from django.utils.functional import cached_property
 
 from nobot.fields import HumanCaptchaField
 
@@ -63,3 +65,22 @@ class ReCaptchaField(HumanCaptchaField):
         'captcha_invalid': _('Incorrect, please try again.'),
         'captcha_error': _('Error verifying input, please try again.'),
     }
+
+
+class ManyToManyDescriptor(related_descriptors.ManyToManyDescriptor):
+
+    @cached_property
+    def related_manager_cls(self):
+        """The constrained_target optimization doesn't play nice with our
+        ManagerBase that has default filtering, so set to None so it's
+        bypassed."""
+        manager = super(ManyToManyDescriptor, self).related_manager_cls
+        setattr(manager, 'constrained_target', None)
+        return manager
+
+
+class ManyToManyField(models.ManyToManyField):
+    def contribute_to_class(self, cls, name, **kwargs):
+        super(ManyToManyField, self).contribute_to_class(cls, name, **kwargs)
+        setattr(cls, self.name, ManyToManyDescriptor(
+            self.remote_field, reverse=False))

--- a/src/olympia/amo/fields.py
+++ b/src/olympia/amo/fields.py
@@ -9,7 +9,7 @@ from django.forms import fields
 from django.utils.translation import ugettext_lazy as _
 from django.utils.functional import cached_property
 
-from nobot.fields import HumanCaptchaField
+from nobot.fields import HumanCaptchaField, HumanCaptchaWidget
 
 
 class PositiveAutoField(models.AutoField):
@@ -59,12 +59,19 @@ class HttpHttpsOnlyURLField(fields.URLField):
         ]
 
 
+class ReCaptchaWidget(HumanCaptchaWidget):
+    """Added to workaround to nobot0.5 not supporting django2.1"""
+    def render(self, name, value, attrs=None, renderer=None):
+        return super(ReCaptchaWidget, self).render(name, value, attrs=attrs)
+
+
 class ReCaptchaField(HumanCaptchaField):
     # Sub-class so we can translate the strings.
     default_error_messages = {
         'captcha_invalid': _('Incorrect, please try again.'),
         'captcha_error': _('Error verifying input, please try again.'),
     }
+    widget_class = ReCaptchaWidget
 
 
 class ManyToManyDescriptor(related_descriptors.ManyToManyDescriptor):

--- a/src/olympia/amo/messages.py
+++ b/src/olympia/amo/messages.py
@@ -20,7 +20,7 @@ message title).
 """
 
 
-class DoubleSafe(safestring.SafeData, jinja2.Markup):
+class DoubleSafe(safestring.SafeText, jinja2.Markup):
     """Double safe all the way: marks safe for django and jinja2.
 
     Even though we're using jinja2 for most of the template rendering, we may
@@ -36,10 +36,11 @@ class DoubleSafe(safestring.SafeData, jinja2.Markup):
 
 def _make_message(title=None, message=None, title_safe=False,
                   message_safe=False):
-    c = {'title': title, 'message': message,
-         'title_safe': title_safe, 'message_safe': message_safe}
-    t = loader.get_template('message_content.html').render(c)
-    return DoubleSafe(t)
+    context = {
+        'title': title, 'message': message,
+        'title_safe': title_safe, 'message_safe': message_safe}
+    tpl = loader.get_template('message_content.html').render(context)
+    return DoubleSafe(tpl)
 
 
 def _is_dupe(msg, request):

--- a/src/olympia/bandwagon/models.py
+++ b/src/olympia/bandwagon/models.py
@@ -11,7 +11,7 @@ from olympia import activity, amo
 from olympia.access import acl
 from olympia.addons.models import Addon
 from olympia.addons.utils import clear_get_featured_ids_cache
-from olympia.amo.fields import PositiveAutoField
+from olympia.amo.fields import ManyToManyField, PositiveAutoField
 from olympia.amo.models import BaseQuerySet, ManagerBase, ModelBase
 from olympia.amo.templatetags.jinja_helpers import absolutify
 from olympia.amo.urlresolvers import reverse
@@ -97,8 +97,8 @@ class Collection(ModelBase):
         default=False,
         help_text='Does this collection only contain Themes?')
 
-    addons = models.ManyToManyField(Addon, through='CollectionAddon',
-                                    related_name='collections')
+    addons = ManyToManyField(
+        Addon, through='CollectionAddon', related_name='collections')
     author = models.ForeignKey(
         UserProfile, null=True, related_name='collections',
         on_delete=models.CASCADE)

--- a/src/olympia/bandwagon/tests/test_views.py
+++ b/src/olympia/bandwagon/tests/test_views.py
@@ -231,6 +231,9 @@ class TestPrivacy(TestCase):
             assert response.status_code == 200
 
 
+@pytest.mark.skipif(
+    django.VERSION[0] >= 2,
+    reason='Legacy UI tests')
 class TestCRUD(TestCase):
     """Test the collection form."""
     fixtures = ('base/users', 'base/addon_3615', 'base/collections')

--- a/src/olympia/devhub/forms.py
+++ b/src/olympia/devhub/forms.py
@@ -12,7 +12,6 @@ from django.forms.widgets import RadioSelect
 from django.utils.safestring import mark_safe
 from django.utils.translation import ugettext, ugettext_lazy as _
 
-import jinja2
 import six
 import waffle
 
@@ -25,6 +24,7 @@ from olympia.addons.models import (
     Addon, AddonCategory, AddonReviewerFlags, AddonUser, Preview)
 from olympia.amo.fields import HttpHttpsOnlyURLField, ReCaptchaField
 from olympia.amo.forms import AMOModelForm
+from olympia.amo.messages import DoubleSafe
 from olympia.amo.templatetags.jinja_helpers import mark_safe_lazy
 from olympia.amo.urlresolvers import reverse
 from olympia.applications.models import AppVersion
@@ -561,7 +561,7 @@ class NewUploadForm(forms.Form):
                     elif version.unreviewed_files:
                         next_url = reverse('devhub.submit.version.details',
                                            args=[self.addon.slug, version.pk])
-                        msg = jinja2.Markup('%s <a href="%s">%s</a>' % (
+                        msg = DoubleSafe('%s <a href="%s">%s</a>' % (
                             ugettext(u'Version {version} already exists.'),
                             next_url,
                             ugettext(u'Continue with existing upload instead?')

--- a/src/olympia/devhub/templates/devhub/forms/widgets/compat_app_input_option.html
+++ b/src/olympia/devhub/templates/devhub/forms/widgets/compat_app_input_option.html
@@ -1,1 +1,1 @@
-{% if wrap_label or widget.wrap_label %}<label class="app {{ widget.compat_app_short }}" {% if widget.attrs.id %} for="{{ widget.attrs.id }}"{% endif %}>{% endif %}{% include "django/forms/widgets/input.html" %}{% if wrap_label %} {{ widget.label }}</label>{% endif %}
+{% if wrap_label or widget.wrap_label %}<label class="app {{ widget.compat_app_short }}" {% if widget.attrs.id %} for="{{ widget.attrs.id }}"{% endif %}>{% endif %}{% include "django/forms/widgets/input.html" %}{% if wrap_label or widget.wrap_label %} {{ widget.label }}</label>{% endif %}

--- a/src/olympia/devhub/tests/test_views_edit.py
+++ b/src/olympia/devhub/tests/test_views_edit.py
@@ -883,7 +883,7 @@ class TestEditMedia(BaseTestEdit):
         https://github.com/mozilla/addons-server/issues/8900"""
         doc = pq(self.client.get(self.media_edit_url).content)
 
-        labels = doc('#icons_default li label')
+        labels = doc('#icons_default li a~label')
 
         assert labels.length == 18
 

--- a/src/olympia/reviewers/tests/test_serializers.py
+++ b/src/olympia/reviewers/tests/test_serializers.py
@@ -206,7 +206,8 @@ class TestAddonBrowseVersionSerializer(TestCase):
         )
 
         extract_version_to_git(self.addon.current_version.pk)
-        self.addon.current_version.refresh_from_db()
+        self.addon.current_version.reload()
+        assert self.addon.current_version.release_notes
         self.version = self.addon.current_version
 
     def serialize(self, **extra_context):

--- a/src/olympia/tags/models.py
+++ b/src/olympia/tags/models.py
@@ -3,7 +3,7 @@ from django.urls import NoReverseMatch
 from django.utils.encoding import python_2_unicode_compatible
 
 from olympia import activity, amo
-from olympia.amo.fields import PositiveAutoField
+from olympia.amo.fields import ManyToManyField, PositiveAutoField
 from olympia.amo.models import ManagerBase, ModelBase
 from olympia.amo.urlresolvers import reverse
 
@@ -21,8 +21,8 @@ class Tag(ModelBase):
     tag_text = models.CharField(max_length=128)
     denied = models.BooleanField(default=False)
     restricted = models.BooleanField(default=False)
-    addons = models.ManyToManyField('addons.Addon', through='AddonTag',
-                                    related_name='tags')
+    addons = ManyToManyField(
+        'addons.Addon', through='AddonTag', related_name='tags')
     num_addons = models.IntegerField(default=0)
 
     objects = TagManager()

--- a/src/olympia/translations/fields.py
+++ b/src/olympia/translations/fields.py
@@ -28,10 +28,17 @@ class TranslationDescriptor(related.ForwardManyToOneDescriptor):
         # happen if the field was set or accessed already), it does a db query
         # to follow the foreign key. We expect translations to be set by
         # queryset transforms, so doing a query is the wrong thing here.
-        try:
-            return getattr(instance, self.cache_name)
-        except AttributeError:
-            return None
+        # django2.0+ uses field.get_cached_value rather than cache_name
+        if hasattr(self.field, 'get_cached_value'):
+            try:
+                return self.field.get_cached_value(instance)
+            except KeyError:
+                return None
+        else:
+            try:
+                return getattr(instance, self.cache_name)
+            except AttributeError:
+                return None
 
     def __set__(self, instance, value):
         lang = translation_utils.get_language()

--- a/src/olympia/translations/models.py
+++ b/src/olympia/translations/models.py
@@ -78,6 +78,9 @@ class Translation(ModelBase):
         else:
             return self.localized_string == other
 
+    def __hash__(self):
+        return hash(self.localized_string)
+
     def clean(self):
         if self.localized_string:
             self.localized_string = self.localized_string.strip()

--- a/src/olympia/translations/query.py
+++ b/src/olympia/translations/query.py
@@ -1,4 +1,5 @@
 import itertools
+from collections import OrderedDict
 
 from django.conf import settings
 from django.db import models
@@ -28,7 +29,7 @@ def order_by_translation(qs, fieldname, model=None):
 
     # Doing the manual joins is flying under Django's radar, so we need to make
     # sure the initial alias (the main table) is set up.
-    if not qs.query.tables:
+    if not qs.query.table_map:
         qs.query.get_initial_alias()
 
     # Force two new joins against the translation table, without reusing any
@@ -38,7 +39,8 @@ def order_by_translation(qs, fieldname, model=None):
     # building the query manually, does not detect that an inner join would
     # remove results and happily simplifies the LEFT OUTER JOINs to
     # INNER JOINs)
-    qs.query = qs.query.clone(TranslationQuery)
+    qs.query = qs.query.clone()
+    qs.query.__class__ = TranslationQuery
     t1 = qs.query.join(
         Join(field.remote_field.model._meta.db_table, model._meta.db_table,
              None, LOUTER, field, True),
@@ -65,9 +67,9 @@ class TranslationQuery(models.sql.query.Query):
     translations.
     """
 
-    def clone(self, klass=None, **kwargs):
+    def clone(self):
         # Maintain translation_aliases across clones.
-        c = super(TranslationQuery, self).clone(klass, **kwargs)
+        c = super(TranslationQuery, self).clone()
         c.translation_aliases = self.translation_aliases
         return c
 
@@ -83,10 +85,15 @@ class SQLCompiler(compiler.SQLCompiler):
     def get_from_clause(self):
         # Temporarily remove translation tables from query.tables so Django
         # doesn't create joins against them.
-        old_tables = list(self.query.tables)
+        # self.query.tables was removed in django2.0+
+        old_tables = (
+            list(self.query.tables) if hasattr(self.query, 'tables') else None)
+        old_map = OrderedDict(self.query.alias_map)
         for table in itertools.chain(*self.query.translation_aliases.values()):
-            if table in self.query.tables:
-                self.query.tables.remove(table)
+            if table in self.query.alias_map:
+                if old_tables:
+                    self.query.tables.remove(table)
+                del self.query.alias_map[table]
 
         joins, params = super(SQLCompiler, self).get_from_clause()
 
@@ -103,20 +110,22 @@ class SQLCompiler(compiler.SQLCompiler):
         # Django had in query.tables, but that seems to be ok.
         for field, aliases in self.query.translation_aliases.items():
             t1, t2 = aliases
-            joins.append(self.join_with_locale(t1))
-            joins.append(self.join_with_locale(t2, fallback))
+            joins.append(self.join_with_locale(t1, None, old_map))
+            joins.append(self.join_with_locale(t2, fallback, old_map))
 
-        self.query.tables = old_tables
+        if old_tables:
+            self.query.tables = old_tables
+        self.query.alias_map = old_map
         return joins, params
 
-    def join_with_locale(self, alias, fallback=None):
+    def join_with_locale(self, alias, fallback, alias_map):
         # This is all lifted from the real sql.compiler.get_from_clause(),
         # except for the extra AND clause.  Fun project: fix Django to use Q
         # objects here instead of a bunch of strings.
         qn = self.quote_name_unless_alias
         qn2 = self.connection.ops.quote_name
 
-        join = self.query.alias_map[alias]
+        join = alias_map[alias]
         lhs_col, rhs_col = join.join_cols[0]
         alias_str = (
             '' if join.table_alias == join.table_name

--- a/src/olympia/translations/tests/test_models.py
+++ b/src/olympia/translations/tests/test_models.py
@@ -255,11 +255,10 @@ class TranslationTestCase(TestCase):
         assert sorted([c, a, b]) == [a, b, c]
 
     def test_sorting_en(self):
-        q = TranslatedModel.objects.all()
+        query = TranslatedModel.objects.all()
         expected = [4, 1, 3]
-
-        assert ids(order_by_translation(q, 'name')) == expected
-        assert ids(order_by_translation(q, '-name')) == (
+        assert ids(order_by_translation(query, 'name')) == expected
+        assert ids(order_by_translation(query, '-name')) == (
             list(reversed(expected)))
 
     def test_order_by_translations_query_uses_left_outer_join(self):

--- a/src/olympia/users/tests/test_commands.py
+++ b/src/olympia/users/tests/test_commands.py
@@ -77,9 +77,7 @@ class TestUpdateDeletedUsers(TestCase):
     def test_updates_deleted_metadata(self):
         user = user_factory(fxa_id='foobar', last_login_ip='192.168.1.1')
 
-        call_command(
-            'update_deleted_users',
-            interactive=False)
+        call_command('update_deleted_users')
 
         # Nothing happened, the user wasn't deleted
         user.refresh_from_db()
@@ -92,9 +90,7 @@ class TestUpdateDeletedUsers(TestCase):
         # Now set something and make sure `.delete` get's called again
         user.update(fxa_id='foobar', last_login_ip='192.168.1.1')
 
-        call_command(
-            'update_deleted_users',
-            interactive=False)
+        call_command('update_deleted_users')
 
         user.refresh_from_db()
         assert user.fxa_id is None

--- a/src/olympia/users/tests/test_views.py
+++ b/src/olympia/users/tests/test_views.py
@@ -1,3 +1,9 @@
+from datetime import datetime, timedelta
+
+from django.conf import settings
+from django.utils.encoding import force_text
+
+from dateutil.parser import parse
 from pyquery import PyQuery as pq
 
 from olympia import amo
@@ -115,3 +121,23 @@ class TestUnsubscribe(UserViewBase):
         doc = pq(r.content)
 
         assert doc('#unsubscribe-fail').length == 1
+
+
+class TestSessionLength(UserViewBase):
+
+    def test_session_does_not_expire_quickly(self):
+        """Make sure no one is overriding our settings and making sessions
+        expire at browser session end. See:
+        https://github.com/mozilla/addons-server/issues/1789
+        """
+        self.client.login(email='jbalogh@mozilla.com')
+        r = self.client.get('/', follow=True)
+        cookie = r.cookies[settings.SESSION_COOKIE_NAME]
+
+        # The user's session should be valid for at least four weeks (near a
+        # month).
+        four_weeks_from_now = datetime.now() + timedelta(days=28)
+        expiry = parse(cookie['expires']).replace(tzinfo=None)
+
+        assert cookie.value != ''
+        assert expiry >= four_weeks_from_now

--- a/src/olympia/users/tests/test_views.py
+++ b/src/olympia/users/tests/test_views.py
@@ -53,14 +53,14 @@ class TestUnsubscribe(UserViewBase):
     def test_correct_url_update_notification(self):
         # Make sure the user is subscribed
         perm_setting = email.NOTIFICATIONS_COMBINED[0]
-        un = UserNotification.objects.create(notification_id=perm_setting.id,
-                                             user=self.user,
-                                             enabled=True)
+        un = UserNotification.objects.create(
+            notification_id=perm_setting.id, user=self.user, enabled=True)
 
         # Create a URL
         token, hash = UnsubscribeCode.create(self.user.email)
-        url = reverse('users.unsubscribe', args=[token, hash,
-                                                 perm_setting.short])
+        url = reverse(
+            'users.unsubscribe', args=[
+                force_text(token), hash, perm_setting.short])
 
         # Load the URL
         r = self.client.get(url)
@@ -72,8 +72,8 @@ class TestUnsubscribe(UserViewBase):
         assert doc('#standalone ul li').length == 1
 
         # Make sure the user is unsubscribed
-        un = UserNotification.objects.filter(notification_id=perm_setting.id,
-                                             user=self.user)
+        un = UserNotification.objects.filter(
+            notification_id=perm_setting.id, user=self.user)
         assert un.count() == 1
         assert not un.all()[0].enabled
 
@@ -84,8 +84,9 @@ class TestUnsubscribe(UserViewBase):
         # Create a URL
         perm_setting = email.NOTIFICATIONS_COMBINED[0]
         token, hash = UnsubscribeCode.create(self.user.email)
-        url = reverse('users.unsubscribe', args=[token, hash,
-                                                 perm_setting.short])
+        url = reverse(
+            'users.unsubscribe', args=[
+                force_text(token), hash, perm_setting.short])
 
         # Load the URL
         r = self.client.get(url)
@@ -97,8 +98,8 @@ class TestUnsubscribe(UserViewBase):
         assert doc('#standalone ul li').length == 1
 
         # Make sure the user is unsubscribed
-        un = UserNotification.objects.filter(notification_id=perm_setting.id,
-                                             user=self.user)
+        un = UserNotification.objects.filter(
+            notification_id=perm_setting.id, user=self.user)
         assert un.count() == 1
         assert not un.all()[0].enabled
 
@@ -107,8 +108,9 @@ class TestUnsubscribe(UserViewBase):
         token, hash = UnsubscribeCode.create(self.user.email)
         hash = hash[::-1]  # Reverse the hash, so it's wrong
 
-        url = reverse('users.unsubscribe', args=[token, hash,
-                                                 perm_setting.short])
+        url = reverse(
+            'users.unsubscribe', args=[
+                force_text(token), hash, perm_setting.short])
         r = self.client.get(url)
         doc = pq(r.content)
 

--- a/src/olympia/zadmin/views.py
+++ b/src/olympia/zadmin/views.py
@@ -194,8 +194,16 @@ def general_search(request, app_id, model_id):
     # This is a hideous api, but uses the builtin admin search_fields API.
     # Expecting this to get replaced by ES so soon, that I'm not going to lose
     # too much sleep about it.
-    cl = ChangeList(request, obj.model, [], [], [], [], obj.search_fields, [],
-                    obj.list_max_show_all, limit, [], obj)
+    args = [request, obj.model, [], [], [], [], obj.search_fields, [],
+            obj.list_max_show_all, limit, [], obj]
+    try:
+        # python3.2+ only
+        from inspect import signature
+        if 'sortable_by' in signature(ChangeList.__init__).parameters:
+            args.append('None')  # sortable_by is a django2.1+ addition
+    except ImportError:
+        pass
+    cl = ChangeList(*args)
     qs = cl.get_queryset(request)
     # Override search_fields_response on the ModelAdmin object
     # if you'd like to pass something else back to the front end.

